### DISCUSSION TITLE
Renovate loses ".npmrc" context when trying to regenerate lock files

### DISCUSSION THREAD

**Discussion by elliotleelewis** [renovatebot/renovate#8939] [Request Help]
**What Renovate type, platform and version are you using?**

Self-hosted in Azure DevOps. Using packages from both NPMJS and a private feed in Azure DevOps (Azure Artifacts)

**Describe the bug**

Renovate is trying to create a PR for a private NPM package, but falls over itself when generating the lock-file, and gets a "401 Unauthorized" from the private feed.

![image](https://user-images.githubusercontent.com/17044667/109601500-fa96f600-7aec-11eb-837e-83a5ff38a724.png)

I'm setting the "npmrc" config option via a command-line parameter to the Renovate runner with the following value (of course with the new lines replaced by "\n"):

```
@jabil:registry=https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/
@jabil:always-auth=true
//pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/:username=VssSessionToken
//pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/:email=not-used@example.com
//pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/:_password=${NPM_TOKEN}
```

We use this same ".npmrc" file everywhere across our organization and I know that it works. It even works here to allow Renovate to access the info about the updated package, just not sure why it doesn't want to use this when it generates the lock-file. Maybe its getting overwrote by something?

**Relevant debug logs**

```
DEBUG: File config
       "config": {}

         "hostRules": [
           {"hostType": "github", "domainName": "github.com", "token": "***********"}
         ]
       }

DEBUG: Combined config
       "config": {
         "hostRules": [
           {"hostType": "github", "domainName": "github.com", "token": "***********"}
         ],
         "repositories": ["UIUX Community/styleguide-legacy"],
         "platform": "azure",
         "endpoint": "https://dev.azure.com/jblinnersource/",
         "token": "***********",
         "npmrc": "***********"
       }
```

```
DEBUG: npm.updateDependency(): dependencies.@jabil/ui-styles = ^0.3.0 (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Updating @jabil/ui-styles in package.json (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)


DEBUG: npmrc file found in repository (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Generating package-lock.json for . (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Spawning npm install to create /tmp/renovate/repos/azure/UIUX Community/styleguide-legacy/package-lock.json (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Updating lock file only (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: No node constraint found - using latest (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Executing command (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)
       "command": "npm install --package-lock-only --ignore-scripts --no-audit"

DEBUG: lock file error (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "npm install --package-lock-only --ignore-scripts --no-audit",
         "stdout": "",
         "stderr": "npm ERR! code E401\nnpm ERR! Unable to authenticate, need: Bearer authorization_uri=https://login.windows.net/bc876b21-f134-4c12-a265-8ed26b7f0f3b, Basic realm=\"https://pkgsprodcus1.pkgs.visualstudio.com/\", TFS-Federated\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate/cache/others/npm/_logs/2021-03-02T05_14_04_390Z-debug.log\n",
         "message": "Command failed: npm install --package-lock-only --ignore-scripts --no-audit\nnpm ERR! code E401\nnpm ERR! Unable to authenticate, need: Bearer authorization_uri=https://login.windows.net/bc876b21-f134-4c12-a265-8ed26b7f0f3b, Basic realm=\"https://pkgsprodcus1.pkgs.visualstudio.com/\", TFS-Federated\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate/cache/others/npm/_logs/2021-03-02T05_14_04_390Z-debug.log\n",
         "stack": "Error: Command failed: npm install --package-lock-only --ignore-scripts --no-audit\nnpm ERR! code E401\nnpm ERR! Unable to authenticate, need: Bearer authorization_uri=https://login.windows.net/bc876b21-f134-4c12-a265-8ed26b7f0f3b, Basic realm=\"https://pkgsprodcus1.pkgs.visualstudio.com/\", TFS-Federated\n\nnpm ERR! A complete log of this run can be found in:\nnpm ERR!     /tmp/renovate/cache/others/npm/_logs/2021-03-02T05_14_04_390Z-debug.log\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:315:20)\n    at ChildProcess.EventEmitter.emit (domain.js:467:12)\n    at maybeClose (internal/child_process.js:1048:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
       },
       "type": "npm"

DEBUG: No updated lock files in branch (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Branch timestamp: 2021-02-15T23:15:44.000Z (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: PR is older than 2 hours, raise PR with lock file errors (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: 1 file(s) to commit (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)

DEBUG: Committing files to branch renovate/jabil-ui-styles-0.x (repository=UIUX Community/styleguide-legacy, branch=renovate/jabil-ui-styles-0.x)
```

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [x] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

Really not sure if I have any additional context here, I've tried to give as much detail as I can in the logs. My only theory is that from the debug logs it looks like Renovate is using the `.npmrc` from the branch, rather than the one from my Renovate config.

**Comment by rarkins** [comment]
Do you have any `.npmrc` file within the repo itself?

**Reply by elliotleelewis** [reply]
Yes there is a `.npmrc` in the repo. Its contents are:

```
@jabil:registry=https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/
@jabil:always-auth=true
```

**Reply by rarkins** [reply]
If you configure `npmrc` in your Renovate config then it will overwrite any in your repo. Could that explain the failure?

**Reply by elliotleelewis** [reply]
Oh well, that's kinda the thing. In theory, if that was happening it would be working. The npmrc in the config is just the same as the one in the repo, but with the auth token added (via an NPM_TOKEN environment variable). It looks like the auth token isn't working here when Renovate tries to generate the lock-file, which doesn't make sense to me because it works 5 seconds earlier when it knows that there's an update and generates the PR initially?

I think when it's generating the lock-file, either the environment variable isn't being passed to the npmrc correctly, or it's using the npmrc in the repo, rather than the one in the config? What do you think? 😄

**Reply by elliotleelewis** [reply]
Okay, I've found some more interesting info out about this. For context, the runner on the 3 examples that I will describe has the "npmrc" config option set via that CLI parameter, with the `NPM_TOKEN` being made available via an environment variable.

Basically, there are 3 different scenarios I have on 3 different repos that will help us understand what's going on here.

| Repo | Custom "npmrc" File | Dependencies on Private Feed | Renovate PR Output |
| :-: | :-: | :-: | --- |
| styleguide | ✔ | ✔ | Creates PR but gets `401 Unauthorized` from Private Feed for the lock-file generation |
| wiki | ✔ | ❌ | Creates PR successfully with lock-file generation |
| meta | ❌ | ❌ | Creates PR but gets `Failed to replace env in config: ${NPM_TOKEN}` when doing lock-file generation |

These results are very interesting. The difference between the lock-files in the repo and in the config parameter is that the config parameter has the auth key with "${NPM_TOKEN}" and the one in the repo does not. When we use the lock file without that auth key, the `401 Unauthorized` is the expected behaviour when trying to access packages on that Private Feed, and in fact, developers run into this locally if they haven't configured their user-level ".npmrc" file with the correct credentials. This tells me that when there's a custom ".npmrc" file in the repo, its using that when generating the lock-file instead of the config parameter, and this is confirmed by the expected behaviour in the `wiki` repo too (the PR generated doesn't need authentication from the Private Feed, and the lock-file in the repo doesn't have an auth token that fails at substitution based on the `meta` repo example, so this completes successfully). The third scenario is also very interesting, as when there's no ".npmrc" in the repo, it does seem to correctly use the config parameter, but as I just mentioned it fails at using the environment variable for the `NPM_TOKEN`.

I really think that this is a bug with Renovate. I think there are 2 issues here:

1. The ".npmrc" file in the repo takes priority over the config parameter when generating lock-files.
1. The environment variables are not correctly being substituted into the config parameter's "npmrc" when generating lock-files.

@rarkins I'd be more than willing to jump on a Zoom/Teams call to take you through this and review? Would love to find a solution to this.

Thanks! 😄

**Reply by rarkins** [reply]
It's a requirement to configure `trustLevel=high` in your bot admin config if you want to do environment substitution such as `${NPM_TOKEN}`. Can you confirm that's the case?

Assuming that's so, and you see `Failed to replace env in config`, it means that the `process.env` that is passed to Renovate does not contain `NPM_TOKEN`. i.e. it's a problem that needs to be solved somehow before you call Renovate.

**Reply by viceice** [reply]
A better solution is to configure a hostRule instead of env substitution. Renovate will add those to the repo npmrc before running npm so auth will succeed.

@rarkins we really should deprecate the env substitution. At least for the docs and suggest hostrules there.

**Reply by elliotleelewis** [reply]
> 
> 
> It's a requirement to configure `trustLevel=high` in your bot admin config if you want to do environment substitution such as `${NPM_TOKEN}`. Can you confirm that's the case?
> 
> Assuming that's so, and you see `Failed to replace env in config`, it means that the `process.env` that is passed to Renovate does not contain `NPM_TOKEN`. i.e. it's a problem that needs to be solved somehow before you call Renovate.

Got it! Just added `--trust-level="high"` and now the `meta` repo example is working fine, thanks! However `styleguide` is still giving the same `401 Unauthorized`, so its still using the repo's `.npmrc`.

**Reply by elliotleelewis** [reply]
> 
> 
> A better solution is to configure a hostRule instead of env substitution. Renovate will add those to the repo npmrc before running npm so auth will succeed.
> 
> @rarkins we really should deprecate the env substitution. At least for the docs and suggest hostrules there.

Oh nice, I haven't seen this `hostRule` property section before. I just checked the Renovate docs and there's not much on there for it (could only find this: https://docs.renovatebot.com/configuration-options/#hostrules). Is there an example somewhere I could follow to add this to my `.npmrc`? :)

```
//pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/:username=VssSessionToken
//pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/:email=not-used@example.com
//pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/:_password=${NPM_TOKEN}
```

**Reply by viceice** [reply]
we need a fix for username / password combo. i'll send that in a few minutes.

the host rule should then **one** of those:
```js
{
  hostRules: [
    {
      hostType: 'npm',
      baseUrl: 'https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/',
      username: 'VssSessionToken',
      password: process.env.NPM_TOKEN
    },
    {
      hostType: 'npm',
      hostName: 'pkgs.dev.azure.com',
      username: 'VssSessionToken',
      password: process.env.NPM_TOKEN
    }
  ]
}
```

and your repo `.npmrc`:
```ini
@jabil:registry=https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/
@jabil:always-auth=true
```

**Reply by elliotleelewis** [reply]
Sweet, sounds good! Looks like your PR for that just went through today, so let me revisit this. Thanks. 😄

**Reply by elliotleelewis** [reply]
@viceice Hmmm, not sure this worked perfectly. Just updated my renovate config to include the hostRules property, and looking at the logs, it looks like its being pulled into Renovate correctly:

<img width="855" alt="image" src="https://user-images.githubusercontent.com/17044667/110246777-f62c6c00-7f36-11eb-9836-a973c60cd826.png">

However I got a different 401 Unauthorized from the Azure Artifacts stuff:

<img width="824" alt="image" src="https://user-images.githubusercontent.com/17044667/110246865-50c5c800-7f37-11eb-9351-1ad02d6e0c80.png">

<img width="1007" alt="image" src="https://user-images.githubusercontent.com/17044667/110246843-3ab80780-7f37-11eb-96b5-59e651409aae.png">

~~The really interesting thing was that when I renamed my environment variable from `NPM_TOKEN` to `AZURE_NPM_TOKEN` (and of course updated the reference to it) I also got a 401, but this time a different error message from Azure Artifacts (see no body here):~~

~~<img width="1006" alt="image" src="https://user-images.githubusercontent.com/17044667/110247176-ef065d80-7f38-11eb-85c9-91f8b3c24a8a.png">~~


Any ideas?

EDIT: Scratch the variable name stuff, not sure why but Azure DevOps didn't want to do the variable replacement for `AZURE_NPM_TOKEN`. See here what the command looks like before and after. Weird one this:

<img width="327" alt="image" src="https://user-images.githubusercontent.com/17044667/110248191-de0c1b00-7f3d-11eb-953a-30a912ec847d.png">
<img width="292" alt="image" src="https://user-images.githubusercontent.com/17044667/110248209-eb290a00-7f3d-11eb-96ef-de69ffddbed4.png">

Oh well, disregard that piece I guess. 😅

**Reply by viceice** [reply]
Did you verified that `baseUrl` is a prefix of the logged `pkgUrl`.

**Reply by elliotleelewis** [reply]
Yeah just double-checked. In the debug logs:

`baseUrl` =>
<img width="959" alt="image" src="https://user-images.githubusercontent.com/17044667/110247887-71dce780-7f3c-11eb-93a7-399eb7f5ec9f.png">

`pkgUrl` => 
<img width="1069" alt="image" src="https://user-images.githubusercontent.com/17044667/110247936-a3ee4980-7f3c-11eb-93c5-aa48efba8f5c.png">

**Reply by elliotleelewis** [reply]
@viceice Okay, this is wild... 😅

This actually worked in the weekly "lock file maintenance" PRs, which means it correctly injected the stuff into those .npmrc files when working, but it still has that 401 Unauthorized at the top of the logs. I've pasted the logs of this run here:

https://gist.github.com/elliotleelewis/655ee880efd03b1b6b0af9c96aa19b93

The interesting lines are 365, 454, and 3593. I guess this means that NPM is authenticating correctly, but Renovate isn't when querying that Azure DevOps endpoint?

EDIT: Just to be clear, yes this lock-file maintenance PR did update dependencies on that private feed, see below:

![image](https://user-images.githubusercontent.com/17044667/110352890-7371e280-8004-11eb-88ef-68b4d9e96117.png)

**Reply by viceice** [reply]
ok, found the issue. Will fix it asap.

we only support `token` here:

https://github.com/renovatebot/renovate/blob/74d7691befa635c1220d8569936273bfb950e748/lib/datasource/npm/get.ts#L133

**Reply by viceice** [reply]
ok, i found a `authorization`  header, so it looks like that is rejected

```
2021-03-08T00:11:51.5294223Z          "options": {
2021-03-08T00:11:51.5294718Z            "headers": {
2021-03-08T00:11:51.5295520Z              "user-agent": "https://github.com/renovatebot/renovate",
2021-03-08T00:11:51.5296395Z              "accept-encoding": "gzip, deflate",
2021-03-08T00:11:51.5297015Z              "authorization": "***********",
2021-03-08T00:11:51.5297798Z              "cache-control": "no-cache",
2021-03-08T00:11:51.5299208Z              "accept": "application/json"
2021-03-08T00:11:51.5299919Z            },
2021-03-08T00:11:51.5301383Z            "url": "https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/@jabil%2Fui-styles",
2021-03-08T00:11:51.5302050Z            "username": "",
2021-03-08T00:11:51.5302445Z            "password": "",
2021-03-08T00:11:51.5302849Z            "method": "GET",
2021-03-08T00:11:51.5303229Z            "http2": false
2021-03-08T00:11:51.5303600Z          },
```


can you try to do a simple curl?

```sh
curl -svo /dev/null -u "VssSessionToken:${NPM_TOKEN}" "https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/@jabil%2Fui-styles"
```

**Reply by elliotleelewis** [reply]
@viceice Okay another wild one.... so I tried that `curl` with the Base64-encoded PAT that we usually use as a `NPM_TOKEN` for our `.npmrc` files and I got a "401 Unauthorized", and then I tried it without being Base64-encoded, and it gave me a "200"...

The Azure DevOps instructions that describe the connection instructions to the private feed say that the PAT needs to be Base64 encoded (see below), so this is an interesting one.

![image](https://user-images.githubusercontent.com/17044667/110479852-d8384600-80b3-11eb-9c00-57526120bf29.png)

**Reply by viceice** [reply]
the hostRule shouldn't be base64, renovate will do this for you before writing to npmrc. npm / yarn will then decode it again when loading npmrc

**Reply by elliotleelewis** [reply]
I see I see. Interesting that the lock file generation worked with it Base64 encoded though? Let me update it and try again. 😄

**Reply by elliotleelewis** [reply]
@viceice Okay, I updated the password to be not Base64 encoded (the one that works with `curl`).

I added a step in the pipeline to output the password just before Renovate runs and I can verify that its correct, but Renovate still has a "401 Unauthorized".

![image](https://user-images.githubusercontent.com/17044667/110485179-5f3bed00-80b9-11eb-8708-5091d1e837c5.png)

With the same original TF400813 error.

![image](https://user-images.githubusercontent.com/17044667/110485253-724ebd00-80b9-11eb-844b-abfc8f23f4ac.png)

I just spent a couple of minutes playing around with `curl` to see if I could get the same error to be reproducible, these are my results.

- Doing `curl -u 'VssSessionToken:${NPM_TOKEN}' ...` works 
- Doing `curl -u ':${NPM_TOKEN}' ...` works (notice the colon before where I paste my PAT).
- Doing `curl -u 'RandomStringHere:${NPM_TOKEN}' ...` works.
- Doing `curl ...` gives the exact same "401 Unauthorized" error that Renovate gets (`TF400813: The user '' is not authorized to access this resource.`)

So it looks like the password isn't being injected into that `authorization: Basic ...` request header?

**Reply by viceice** [reply]
can we double check your `renovate.js` please

**Reply by elliotleelewis** [reply]
@viceice So I have these 2 configs here:

The pipeline in Azure DevOps runs with this:

```yaml
  - task: jyc.vsts-extensions-renovate-me.default-build-task.RenovateMe@0
    displayName: 'Renovate'
    inputs:
      renovateOptionsVersion: 'latest'
      renovateOptionsArgs: '--host-rules="[{\"hostType\":\"npm\",\"baseUrl\":\"https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/\",\"username\":\"VssSessionToken\",\"password\":\"$(NPM_TOKEN)\"}]"'
    env:
      LOG_LEVEL: 'debug'
      NPM_TOKEN: ${{ parameters.npmToken }}
      GITHUB_COM_TOKEN: ${{ parameters.githubToken }}
```

Which runs Renovate like so:

`npx renovate@latest UIUX Community/styleguide --platform azure --endpoint https://dev.azure.com/jblinnersource/ --token *** --host-rules=[{"hostType":"npm","baseUrl":"https://pkgs.dev.azure.com/jblinnersource/ab32dfdb-775d-44da-a767-c4e669a46d63/_packaging/uiux-community/npm/registry/","username":"VssSessionToken","password":"***"}]`

And then in the repo that I'm running this against I have this `renovate.json`:

```json
{
  "extends": ["config:base", ":preserveSemverRanges"],
  "lockFileMaintenance": {
    "enabled": true
  },
  "prConcurrentLimit": 0,
  "prHourlyLimit": 0
}
```

**Reply by viceice** [reply]
I know the issue, your `NPM_TOKEN` is also used by renovate and forces bearer auth. can you rename it to something else please, like `AZURE_NPM_TOKEN`

https://github.com/renovatebot/renovate/blob/606dd318816150805bd866c5678f14b7d8229dc4/lib/datasource/npm/get.ts#L121-L127

**Reply by elliotleelewis** [reply]
Okay, I see, that makes sense. So I just updated that, and now we're back to the start. 😅 

The PRs for those private dependencies get generated, but the lock-files for those PRs are failed to be generated.

![image](https://user-images.githubusercontent.com/17044667/110502092-df1d8380-80c8-11eb-82d2-90dc41054366.png)

![image](https://user-images.githubusercontent.com/17044667/110502278-0ecc8b80-80c9-11eb-82e8-76554e643bee.png)

Are you sure that the host rules password is Base64 encoded before its injected into that `.npmrc`? Because it's weird that the lock-file generation worked with the pre-encoded PAT in the hostRules, but now that its not pre-encoded, its not? 🤔

**Reply by viceice** [reply]
I'll check this

**Reply by viceice** [reply]
My bad, forgot to do base64 encoding in #8967. fixing asap

**Reply by viceice** [reply]
now we need to wait for #9051 to get live in app. Sorry

**Reply by elliotleelewis** [reply]
@viceice Nice spot!! Thank you for the quick fixes, really appreciate all the help! 😄

**Reply by elliotleelewis** [reply]
@viceice It works!! 🎉

Just tested out the new release and it added the lock-files to the PRs that were pending for our custom dependencies. Thank you so much for all your help!

**Reply by viceice** [reply]
Thanks for confirming
